File: bubble_sort/perl_bubble_sort/bubble_sort.pl
#!/usr/bin/perl -w

use strict;
use warnings;

sub main {
  my @arr = (10, 2, 4, 1, 6, 5, 8, 7, 3, 9);
  if( $#ARGV+1 != 0 ) {
    print STDERR "Usage: $0 <noargs>\n";
    exit(1);
  }
  &bubble_sort(@arr);
  #print $ARGV[0], "\n";
  exit(0);
}

&main();

sub bubble_sort {
  my $array = shift;
  my $not_complete = 1;
  my $index;
  my $len = ((scalar @$array) - 2);
  while ($not_complete) {
    $not_complete = 0;
    foreach $index (0 .. $len) {
      if (@$array[$index] > @$array[$index + 1]) {
        my $temp = @$array[$index + 1];
        @$array[$index + 1] = @$array[$index];
        @$array[$index] = $temp;
        $not_complete = 1;
      }
    }
  }
}

File: bubble_sort/ruby_bubble_sort/bubble_sort.rb
def main()
  if ARGV.length != 0 then
    puts "Usage: #{$0} <noarg>"
    exit(1)
  end
  puts ARGV[0]

  list = 6, 3, 10, 1, 8, 9, 7, 4, 2, 5
  puts bubble_sort(list)
end

def bubble_sort(list)

  slist = list.clone

  for i in 0..(slist.length - 1)
    for j in 0..(slist.length - i - 2)
        if ( slist[j + 1] <=> slist[j] ) == -1
            slist[j], slist[j + 1] = slist[j + 1], slist[j]
        end
    end
  end

  return slist
end

main()
